### PR TITLE
[#326] [Part 1/2] Add Kover to template 

### LINKED ISSUE #326: Update code coverage tool to Kover instead of Jacoco
## Why

Jacoco doesn't work correctly when [updating to the latest versions of Gradle & Java](https://github.com/nimblehq/android-templates/pull/323)

[Kover](https://github.com/Kotlin/kotlinx-kover) will not only fix this, but it's also a lot more convenient to set up and to work with (good for the future)

## Who Benefits?

Developers, this is essential to provide a good developer experience

### PR DESCRIPTION
**Pull request by ryan-conway**
#326

### Part 1
- [x] Added Kover to the template
- [x] Updated github workflows
- [x] Remove Jacoco references from `buildSrc`

### Part 2
- [ ] Remove Jacoco references from `sample-xml` and `sample-compose`

## Insight 📝

Needed to update the `DangerFile` and `Gemfile` to get CI to work properly. Will update the sample projects in a separate PR once the template is all good

## Proof Of Work 📹

image

### CODE REVIEW

File: .cicdtemplate/.github/workflows/deploy_staging_and_production_to_firebase_app_distribution.yml [hunk 1]
```diff
@@ -46,16 +46,15 @@ jobs:
...
+        run: ./gradlew koverMergedHtmlReport
 
       - name: Archive code coverage reports
         uses: actions/upload-artifact@v2
         with:
           name: CodeCoverageReports
           path: |
-            app/build/reports/jacoco/jacocoTestReport/
-            data/build/reports/jacoco/jacocoTestReport/
+            build/reports/kover/merged/
```
Comment (Tuubz) [inline .cicdtemplate/.github/workflows/deploy_staging_and_production_to_firebase_app_distribution.yml:57]: Could be minimised to `path: build/reports/kover/merged/`

Same goes for `run_detekt_and_unit_tests.yml` 🚀

File: template/app/build.gradle.kts [hunk 2]
```diff
@@ -110,6 +110,13 @@ android {
             isIncludeAndroidResources = true
             isReturnDefaultValues = true
         }
+        unitTests.all {
+            if (it.name != "testStagingDebugUnitTest") {
+                it.extensions.configure(kotlinx.kover.api.KoverTaskExtension::class) {
+                    isDisabled.set(true)
+                }
+            }
+        }
```
Comment (Tuubz) [inline template/app/build.gradle.kts:119]: Should we add this in the `data` module too? 💭

File: template/buildSrc/src/main/java/Versions.kt [hunk 3]
```diff
@@ -25,11 +25,11 @@ object Versions {
     const val HILT_VERSION = "2.38.1"
 
     const val JAVAX_INJECT_VERSION = "1"
-    const val JACOCO_VERSION = "0.8.7"
```
Comment (Tuubz) [inline template/buildSrc/src/main/java/Versions.kt:28]: Let's also remove `jacoco.exec` in `.gitignore` ⭐

File: template/app/build.gradle.kts (hunk 2 above)
Comment (Tuubz) [reply template/app/build.gradle.kts:119]: Also, on [this line](https://github.com/nimblehq/android-templates/pull/344/files#diff-8d83febc7bad9355f66dcdf21f1a25e2b5ee979000e8b11b4446d880542f969bR59-R63), there's a comment regarding Jacoco. 

Would you mind re-checking this? 🦖

File: .cicdtemplate/.github/workflows/deploy_staging_and_production_to_firebase_app_distribution.yml [hunk 4]
```diff
@@ -46,16 +46,15 @@ jobs:
           name: DetektReports
           path: build/reports/detekt/
 
-      - name: Run unit tests and Jacoco
-        run: ./gradlew jacocoTestReport
+      - name: Run unit tests with Kover
+        run: ./gradlew koverMergedHtmlReport
```
Comment (Tuubz) [inline .cicdtemplate/.github/workflows/deploy_staging_and_production_to_firebase_app_distribution.yml:50]: Should we be calling `koverMergedReport` instead?

Then we can archive the HTML version, and still use XML version for Danger 👯

**Review by Tuubz — COMMENTED** [review]
🔥 🔥 🔥

File: .cicdtemplate/.github/workflows/deploy_staging_and_production_to_firebase_app_distribution.yml (hunk 4 above)
Comment (Tuubz) [reply .cicdtemplate/.github/workflows/deploy_staging_and_production_to_firebase_app_distribution.yml:50]: @ryan-conway My bad... This one is correct already 🙇  

But for `run_detekt_and_unit_tests.yml` it should be `koverMergedHtmlReport` if I'm not mistaken.

File: .github/workflows/run_detekt_and_unit_tests.yml [hunk 5]
```diff
@@ -37,12 +37,12 @@ jobs:
           name: DetektReportsTemplate
           path: template/build/reports/detekt/
 
-      - name: Run unit tests and Jacoco on template
+      - name: Run unit tests with Kover on template
         working-directory: ./template
-        run: ./gradlew jacocoTestReport
+        run: ./gradlew koverMergedXmlReport
```
Comment (Tuubz) [inline .github/workflows/run_detekt_and_unit_tests.yml:42]: Woops, this should be `koverMergedHtmllReport` too 🙇 

Side question: Will you make the changes for sample in a separate pull request?

Comment (Tuubz) [reply .github/workflows/run_detekt_and_unit_tests.yml]: Sounds good, please don't forget to update the PR description 😁

File: template/data/build.gradle.kts [hunk 6]
```diff
@@ -52,6 +45,16 @@ android {
         xmlReport = true
         xmlOutput = file("build/reports/lint/lint-result.xml")
     }
+
+    testOptions {
+        unitTests.all {
+            if (it.name != "testStagingDebugUnitTest") {
```
Comment (Tuubz) [inline template/data/build.gradle.kts:51]: @ryan-conway Otherwise the tests for `:data` won't run 

```suggestion
            if (it.name != "testDebugUnitTest") {
```

Last remark, could you please add a PoW to the PR description? 🙏

**Review by Tuubz — APPROVED** [review]
💪

**Comment by chornerman** [comment]
@ryan-conway I found some Jacoco refs inside `new_project.kts` and `bitrise.yml`. Do we have to update these files as well? 🤔

File: template/app/build.gradle.kts [hunk 7]
```diff
@@ -55,13 +55,6 @@ android {
...
             signingConfig = signingConfigs[BuildType.DEBUG]
             buildConfigField("String", "BASE_API_URL", "\"https://jsonplaceholder.typicode.com/\"")
-            /**
-             * From AGP 4.2.0, Jacoco generates the report incorrectly, and the report is missing
-             * some code coverage from module. On the new version of Gradle, they introduce a new
-             * flag [testCoverageEnabled], we must enable this flag if using Jacoco to capture
-             * coverage and creates a report in the build directory.
-             * Reference: https://developer.android.com/reference/tools/gradle-api/7.1/com/android/build/api/dsl/BuildType#istestcoverageenabled
-             */
             isTestCoverageEnabled = true
```
Comment (kaungkhantsoe) [inline template/app/build.gradle.kts:58]: Do we still need this option for KOver? I don't think we need it anymore. No?

File: Gemfile.lock [hunk 8]
```diff
@@ -1,19 +1,19 @@
 GEM
   remote: https://rubygems.org/
   specs:
-    addressable (2.8.0)
-      public_suffix (>= 2.0.2, < 5.0)
+    addressable (2.8.1)
```
Comment (lydiasama) [inline Gemfile.lock:4]: @kaungkhantsoe Do we need to update these versions to install KOver? Can we separate it to another PR?

Comment (kaungkhantsoe) [reply Gemfile.lock:4]: @lydiasama I think Ryan reinitiated the gem. I am using `2.8.0` in my IC.

**Comment by ryan-conway** [comment]
@chornerman Updated those files, but not too sure if `bitrise.yml` is correctly done. Mind checking? 😄

**Comment by chornerman** [comment]
> 

Awesome, will check it out 🚀

File: Gemfile.lock (hunk 8 above)
Comment (lydiasama) [reply Gemfile.lock:4]: Oh, I think I tagged the wrong one. @kaungkhantsoe  @ryan-conway  Sorry. 🙇🏻‍♀️

I see. Thanks for the explanation. 👍🏻

**Review by kaungkhantsoe — APPROVED** [review]

**Review by lydiasama — APPROVED** [review]
I can't find the vestige for jacoco in the template now. Nice job. 👍🏻

**Review by AVI5HEK — APPROVED** [review]

**Review by chornerman — APPROVED** [review]
LGTM 🚀
